File: Leetcode/Python/_69.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 1:
            return x
        elif x < 4:
            return 1
        elif x < 9:
            return 2
        elif x < 16:
            return 3
        left = 0
        right = x // 3
        while left < right:
            middle = (left + right) // 2
            if middle * middle == x:
                return middle
            elif 0 < x - middle * middle < 2 * middle + 1:
                return middle
            elif middle * middle > x:
                right = middle
            else:
                left = middle + 1
                

File: Leetcode/cpp/69.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <algorithm>

class Solution {
public:
    int mySqrt(int x) {
        if (x < 2) {
            return x;
        }

        int left = 2, right = x / 2;
        while (left <= right) {
            long long mid = left + (right - left) / 2;
            if (mid * mid == x) {
                return mid;
            } else if (mid * mid < x) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right;
    }
};
int main() {
    Solution sol;
	int input = 8;
	int expected_output = 2;
	
	int output = sol.mySqrt(input);
	assert (expected_output == output && "Test #1 failed!");
	std::cout << "Tests are passed!" << std::endl;

}
